examples/knative/go/kn-go-tagging/tagging/tagging_test.go

```diff
@@ -0,0 +1,396 @@
+package tagging
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/cloudevents/sdk-go/v2/event"
+	"github.com/vmware/govmomi/simulator"
+	_ "github.com/vmware/govmomi/vapi/simulator"
+	"github.com/vmware/govmomi/vapi/tags"
+	"github.com/vmware/govmomi/vim25"
+	"github.com/vmware/govmomi/vim25/mo"
+	"github.com/vmware/govmomi/vim25/types"
+	"go.uber.org/zap/zaptest"
+	"knative.dev/pkg/logging"
+)
+
+func TestParseEventMoRef(t *testing.T) {
+	type vm struct {
+		vmType string
+		Value  string
+	}
+
+	var tests = []struct {
+		name      string
+		path      string
+		expectErr bool
+		want      *vm
+	}{
+		{
+			"Test that event is readable",
+			"testdata/event.json",
+			false,
+			&vm{vmType: "VirtualMachine", Value: "vm-10000"},
+		},
+		{
+			"Event should return error if vm type and value are null",
+			"testdata/eventErr1.json",
+			true,
+			nil,
+		},
+		{
+			"Event should return error if vm info is null",
+			"testdata/eventErr2.json",
+			true,
+			nil,
+		},
+		{
+			"Event should return error if vm parent is null",
+			"testdata/eventErr3.json",
+			true,
+			nil,
+		},
+		{
+			"Event should return error if data is null",
+			"testdata/eventErr4.json",
+			true,
+			nil,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			e := jsonFileToEvent(tt.path)
+			moRef, err := parseEventMoRef(e)
+
+			if tt.expectErr && err == nil {
+				t.Error("expected error, but got none")
+			}
+
+			if !tt.expectErr && err != nil {
+				if moRef.Type != tt.want.vmType {
+					t.Errorf("got: %q, want: %q", moRef.Type, tt.want.vmType)
+				}
+
+				if moRef.Value != tt.want.Value {
+					t.Errorf("got: %q, want: %q", moRef.Value, tt.want.Value)
+				}
+			}
+		})
+	}
+}
+
+func TestRun(t *testing.T) {
+	const (
+		username = "administrator@vsphere.local"
+		password = "pass"
+	)
+
+	secretsDir := createSecret(t, username, password)
+	defaultEnv := EnvConfig{
+		Port:       50001,
+		SecretPath: secretsDir,
+		Insecure:   true, // vcsim
+	}
+
+	t.Run("fail with authentication error", func(t *testing.T) {
+		model := simulator.VPX()
+		defer model.Remove()
+		if err := model.Create(); err != nil {
```

Comment:
nit: all these `err` checks can be simplified with `assert.NilError()` from https://pkg.go.dev/github.com/vieux/docker@v0.6.3-0.20161004191708-e097c2a938c7/pkg/testutil/assert#NilError

Reply:
I don't like to have to remember how yet another person's library works and their particular methods just to save a couple lines of code, but I'll make the changes for this test. I agree it does look cleaner. One fine detail I don't know (and don't feel like looking up) is how does this `assert` library differentiate between what I used to do with `t.Fatal` and `t.Error`. I'm sure there are other things I'd have to look up and this overhead of learning is a burden on me as well as developers who follow.

Reply:
It's weird. I guess I don't know how to import that library...

```
go mod tidy
go: finding module for package github.com/vieux/docker/pkg/testutil/assert
github.com/vmware-samples/vcenter-event-broker-appliance/examples/knative/go/kn-go-tagging/tagging tested by
        github.com/vmware-samples/vcenter-event-broker-appliance/examples/knative/go/kn-go-tagging/tagging.test imports
        github.com/vieux/docker/pkg/testutil/assert: module github.com/vieux/docker@latest found (v0.6.2), but does not contain package github.com/vieux/docker/pkg/testutil/assert
```

Reply:
I'll fix those after this one is merge, don't worry.

Reply:
I'm feeling the time crunch due to wanting to get things in before the holiday break, so much appreciated, @embano1!